docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+
+**Design principle:** All communication required to implement and monitor the
+consensus network should use P2P, including the various synchronizations.
+
+### Options for ABCI Transport
+
+The majority of current usage is in Go, and the majority of that is mediated by
+the Cosmos SDK, which uses the "direct call" interface. There is probably some
+opportunity to clean up the implementation of that code, notably by inverting
+which interface is at the "top" of the abstraction stack (currently it acts
+like an RPC interface, and escape-hatches into the direct call). However, this
+general approach works fine and doesn't need to be fundamentally changed.
+
+For applications _not_ written in Go, the two remaining options are the
+"socket" protocol (another variation on varint-prefiexed protobuf messages over
+an unstructured stream) and gRPC. It would be nice if we could get rid of one
+of these to reduce (unneeded?) optionality, and the "socket" protocol is the
+most obvious choice. gRPC is more complex, but is already widely used in the
+rest of the ecosystem (including the Cosmos SDK).
+
+The main question is whether requiring gRPC imposes an undue burden on the
+authors of applications in languages without good gRPC support. I propose that
+this is not a large constituency -- and if some important use case does arise
+later, it would not be too difficult for that application author to write a
+little proxy (in Go) that bridges the convenient SDK APIs into a simpler
+protocol.
+
+**Design principle:** It is better for an uncommon special case to carry the
+burdens of its specialness, than to bake an escape hatch into the infrastructure.
+
+**Recommendation:** We should deprecate and remove the socket protocol.
+
+### Options for RPC Transport
+
+[ADR 057][adr-57] proposes using gRPC for the Tendermint RPC implementation.
+This is still possible, but if we are able to simplify and decouple the
+concerns as described above, I do not think it should be necessary.
+
+While JSON-RPC is not the best possible RPC protocol for all situations, it has
+some advantages over gRPC for our domain. Specifically:
+
+- It is easy to call JSON-RPC manually from the command-line, which helps with
+  a common concern for the RPC service, local debugging and operations.
+
+  Relatedly: JSON is relatively easy for humans to read and write, and it can
+  be easily copied and pasted to share sample queries and debugging results in
+  chat, issue comments, and so on. Ideally, the RPC service will not be used
+  for activities where the costs of a text protocol are important compared to
+  its legibility and manual usability benefits.
+
+- gRPC has an enormous dependency footprint for both clients and servers, and
+  many of the features it provides to support security and performance
+  (encryption, compression, streaming, etc.) are mostly irrelevant to local
+  use. Tendermint already needs to include a gRPC client for the remote signer,
```

Comment:
Are we assuming that the RPC is predominantly designed for local use?

Reply:
I'm proposing that we try to get to that point, yes. Right now that seems _not_ to be the case, I think mainly because of light clients (if not ours then IBC?)

Reply:
Does that mean that you would you want the event system to not be accessed via the RPC?

Reply:
> Does that mean that you would you want the event system to not be accessed via the RPC?

I don't have a strong preference about whether events are visible to the RPC service, except that I would like everything to keep working "in production" if the node operator were to firewall or disable the RPC endpoint. If being able to subscribe to events is useful for local debugging, I don't see a strong reason to _remove_ it from the RPC service—my hope is to get RPC off the critical path for the consensus network.

So if the event system COULD be removed from the RPC service, I'd be happy, but I don't actually care if it IS removed.